src/meta/src/hummock/compaction/picker/partition_intral_sub_level_picker.rs

```diff
@@ -0,0 +1,540 @@
+    ) -> Option<CompactionInput> {
+        assert!(levels.can_partition_by_vnode());
+        let l0 = levels.l0.as_ref().unwrap();
+        let max_sub_level_id = self
+            .partitions
+            .iter()
+            .map(|partition| {
+                partition
+                    .sub_levels
+                    .last()
+                    .map(|level| level.sub_level_id)
+                    .unwrap_or(u64::MAX)
+            })
+            .min()
+            .unwrap_or(0);
+
+        for (idx, level) in l0.sub_levels.iter().enumerate() {
+            if level.level_type() != LevelType::Nonoverlapping
+                || level_handlers[0].is_level_pending_compact(level)
+            {
+                continue;
+            }
+
+            if levels.vnode_partition_count == 0
+                && level.total_file_size > self.config.sub_level_max_compaction_bytes / 2
+            {
+                continue;
+            }
+
+            if level.vnode_partition_count > 0 && level.sub_level_id <= max_sub_level_id {
+                continue;
+            }
+
+            let max_compaction_bytes = std::cmp::min(
+                self.config.max_compaction_bytes,
+                self.config.max_bytes_for_level_base,
+            );
+
+            let mut compaction_bytes = 0;
+            let mut compaction_file_count = 0;
+            let mut input_levels = vec![];
+            if l0.sub_levels.len() - idx < self.config.level0_sub_level_compact_level_count as usize
+            {
+                continue;
+            }
+
+            let mut wait_enough = true;
+            for next_level in l0.sub_levels.iter().skip(idx) {
+                if compaction_file_count >= self.config.level0_max_compact_file_number
+                    || compaction_bytes >= max_compaction_bytes
+                {
+                    break;
+                }
+
+                if level.vnode_partition_count == 0 && next_level.vnode_partition_count > 0 {
+                    break;
+                }
+
+                if level_handlers[0].is_level_pending_compact(next_level) {
+                    wait_enough = false;
+                    break;
+                }
+
+                compaction_bytes += next_level.total_file_size;
+                compaction_file_count += next_level.table_infos.len() as u64;
+
+                input_levels.push(InputLevel {
+                    level_idx: 0,
+                    level_type: next_level.level_type,
+                    table_infos: next_level.table_infos.clone(),
+                });
+            }
+
+            if input_levels.len() < self.config.level0_sub_level_compact_level_count as usize
+                && (!wait_enough || level.vnode_partition_count > 0)
+            {
+                continue;
+            }
+
+            input_levels.reverse();
+
+            let vnode_partition_count =
+                if compaction_bytes >= self.config.sub_level_max_compaction_bytes || wait_enough {
```

Comment:
I think it is not a good idea, because if the `sub_level_max_compaction_bytes ` increase, the avg count of input level compacting to base-level may decrease.